ferminet/loss.py

```diff
@@ -255,3 +256,133 @@ def total_energy_jvp(primals, tangents):  # pylint: disable=unused-variable
+    loss is averaged over the batch and over all devices inside a pmap.
+  """
+  batch_local_energy = jax.vmap(
+      local_energy,
+      in_axes=(
+          None,
+          0,
+          networks.FermiNetData(positions=0, spins=0, atoms=0, charges=0),
+      ),
+      out_axes=0,
+  )
+  batch_network = jax.vmap(network, in_axes=(None, 0, 0, 0, 0), out_axes=0)
+
+  @jax.custom_jvp
+  def total_energy(
+      params: networks.ParamTree,
+      key: chex.PRNGKey,
+      data: networks.FermiNetData,
+  ) -> Tuple[jnp.ndarray, AuxiliaryLossData]:
+    """Evaluates the total energy of the network for a batch of configurations.
+
+    Note: the signature of this function is fixed to match that expected by
+    kfac_jax.optimizer.Optimizer with value_func_has_rng=True and
+    value_func_has_aux=True.
+
+    Args:
+      params: parameters to pass to the network.
+      key: PRNG state.
+      data: Batched MCMC configurations to pass to the local energy function.
+
+    Returns:
+      (loss, aux_data), where loss is the mean energy, and aux_data is an
+      AuxiliaryLossData object containing the variance of the energy and the
+      local energy per MCMC configuration. The loss and variance are averaged
+      over the batch and over all devices inside a pmap.
+    """
+    keys = jax.random.split(key, num=data.positions.shape[0])
+    e_l = batch_local_energy(params, keys, data)
+    loss = constants.pmean(jnp.mean(e_l))
+    loss_diff = e_l - loss
+    variance = constants.pmean(jnp.mean(loss_diff * jnp.conj(loss_diff)))
+
+    def batch_local_energy_pos(pos):
+      network_data = networks.FermiNetData(positions=pos, spins=data.spins, 
+          atoms=data.atoms, charges=data.charges)
+      return batch_local_energy(params, keys, network_data).sum()
+    
+    grad_e_l = jax.grad(batch_local_energy_pos)(data.positions)
+    grad_e_l = jnp.tanh(jax.lax.stop_gradient(grad_e_l))
+    return loss, AuxiliaryLossData(variance=variance.real, local_energy=e_l, 
+        clipped_energy=e_l, grad_local_energy=grad_e_l)
+
+  @total_energy.defjvp
+  def total_energy_jvp(primals, tangents):  # pylint: disable=unused-variable
+    """Custom Jacobian-vector product for unbiased local energy gradients."""
+    params, key, data = primals
+    loss, aux_data = total_energy(params, key, data)
+
+    if clip_local_energy > 0.0:
+      aux_data.clipped_energy, diff = clip_local_values(
+          aux_data.local_energy,
+          loss,
+          clip_local_energy,
+          clip_from_median,
+          center_at_clipped_energy,
+          complex_output)
+    else:
+      diff = aux_data.local_energy - loss
+
+    def log_q(_params, _pos, _spins, _atoms, _charges):
+      out = batch_network(_params, _pos, _spins, _atoms, _charges)
+      kfac_jax.register_normal_predictive_distribution(out[:,None])
+      return out.sum()
+
+    score = jax.grad(log_q, argnums=1)
+    primals = (primals[0], data.positions, data.spins, data.atoms, data.charges)
```

Comment:
Clearer and more robust (e.g. if we ever change the args order to `total_energy`) to pass in `params` instead of primals[0] here